File: client/mobile/ios/SecurePerimeter/NoiseClient.swift
//
//  NoiseClient.swift
//  SecurePerimeter
//
//  Noise Protocol XX handshake client with attestation verification.
//  This is a stub demonstrating the API surface - actual implementation
//  requires a Noise Protocol library (e.g., swift-noise-protocol).
//

import Foundation
import Network

// MARK: - Types

/// Connection state for the Noise handshake
public enum NoiseConnectionState {
    case disconnected
    case connecting
    case handshaking
    case connected
    case failed(Error)
}

/// Result of attestation verification
public struct AttestationResult {
    public let valid: Bool
    public let attestationType: AttestationType
    public let errors: [String]
    public let warnings: [String]
    
    public enum AttestationType {
        case sevSnp
        case tdx
        case unknown
    }
}

/// Configuration for NoiseClient connection
public struct NoiseClientConfig {
    /// Server hostname
    public let host: String
    /// Server port
    public let port: UInt16
    /// Use TLS as transport layer
    public let useTls: Bool
    /// Expected image digest for Rekor lookup
    public let expectedImageDigest: String?
    /// Skip attestation verification (DANGEROUS - testing only)
    public let skipAttestation: Bool
    /// Connection timeout in seconds
    public let timeoutSeconds: TimeInterval
    
    public init(
        host: String,
        port: UInt16,
        useTls: Bool = true,
        expectedImageDigest: String? = nil,
        skipAttestation: Bool = false,
        timeoutSeconds: TimeInterval = 30.0
    ) {
        self.host = host
        self.port = port
        self.useTls = useTls
        self.expectedImageDigest = expectedImageDigest
        self.skipAttestation = skipAttestation
        self.timeoutSeconds = timeoutSeconds
    }
}

/// Encrypted session after successful handshake
public protocol NoiseSession {
    /// Send encrypted data to the server
    func send(_ data: Data) throws
    
    /// Receive handler - called when encrypted data arrives
    var onData: ((Data) -> Void)? { get set }
    
    /// Close the session
    func close()
    
    /// Server's static public key (32 bytes for X25519)
    var remotePublicKey: Data { get }
    
    /// Handshake hash for channel binding (32 bytes)
    var handshakeHash: Data { get }
    
    /// Attestation verification result
    var attestation: AttestationResult { get }
}

// MARK: - NoiseClient

/// Noise Protocol client with hardware attestation verification
///
/// ## Overview
/// `NoiseClient` establishes encrypted connections to secure perimeter servers
/// running in Trusted Execution Environments (TEEs). It performs:
///
/// 1. TCP/TLS connection to the server
/// 2. Noise XX handshake with ephemeral and static key exchange
/// 3. Attestation quote extraction from handshake payload
/// 4. Hardware signature verification against AMD/Intel root CA
/// 5. Measurement comparison against Rekor transparency log
///
/// ## Usage
/// ```swift
/// let config = NoiseClientConfig(
///     host: "server.example.com",
///     port: 9000,
///     expectedImageDigest: "sha256:abc123..."
/// )
///
/// let client = NoiseClient(config: config)
/// let session = try await client.connect()
///
/// // Verify attestation passed
/// guard session.attestation.valid else {
///     print("Attestation failed: \(session.attestation.errors)")
///     return
/// }
///
/// // Send encrypted message
/// try session.send("Hello, secure server!".data(using: .utf8)!)
///
/// // Receive messages
/// session.onData = { data in
///     print("Received: \(String(data: data, encoding: .utf8)!)")
/// }
/// ```
///
/// ## Dependencies
/// - swift-noise-protocol or similar Noise implementation
/// - CryptoKit for X25519 key exchange
/// - Network.framework for TCP/TLS connections
///
public class NoiseClient: ObservableObject {
    
    // MARK: - Properties
    
    private let config: NoiseClientConfig
    private var connection: NWConnection?
    
    @Published public private(set) var state: NoiseConnectionState = .disconnected
    
    // Noise handshake state (stub - actual implementation needs noise-protocol lib)
    private var localStaticKeypair: (publicKey: Data, secretKey: Data)?
    private var localEphemeralKeypair: (publicKey: Data, secretKey: Data)?
    private var remoteStaticPublicKey: Data?
    private var remoteEphemeralPublicKey: Data?
    
    // MARK: - Init
    
    public init(config: NoiseClientConfig) {
        self.config = config
        self.localStaticKeypair = generateKeypair()
    }
    
    // MARK: - Connection
    
    /// Connect to the server and perform attested Noise handshake
    ///
    /// - Returns: An encrypted session if attestation verification passes
    /// - Throws: `NoiseError` if connection or attestation fails
    public func connect() async throws -> NoiseSession {
        state = .connecting
        
        // Step 1: Fetch expected measurements from Rekor if image digest provided
        var expectedMeasurements: ExpectedMeasurements?
        if let digest = config.expectedImageDigest {
            expectedMeasurements = try await fetchMeasurementsFromRekor(imageDigest: digest)
        }
        
        // Step 2: Establish TCP/TLS connection
        let endpoint = NWEndpoint.hostPort(
            host: NWEndpoint.Host(config.host),
            port: NWEndpoint.Port(rawValue: config.port)!
        )
        
        let parameters: NWParameters = config.useTls ? .tls : .tcp
        connection = NWConnection(to: endpoint, using: parameters)
        
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            connection?.stateUpdateHandler = { [weak self] state in
                switch state {
                case .ready:
                    continuation.resume()
                case .failed(let error):
                    self?.state = .failed(error)
                    continuation.resume(throwing: error)
                default:
                    break
                }
            }
            connection?.start(queue: .global())
        }
        
        state = .handshaking
        
        // Step 3: Perform Noise XX handshake
        // XX pattern:
        // -> e                    (send ephemeral public key)
        // <- e, ee, s, es         (receive server ephemeral + static + attestation)
        // -> s, se                (send our static public key)
        
        let session = try await performHandshake(expectedMeasurements: expectedMeasurements)
        
        state = .connected
        return session
    }
    
    // MARK: - Private Methods
    
    private func performHandshake(expectedMeasurements: ExpectedMeasurements?) async throws -> NoiseSession {
        guard let connection = connection else {
            throw NoiseError.notConnected
        }
        
        // Generate ephemeral keypair for this handshake
        localEphemeralKeypair = generateKeypair()
        
        // Step 1: Send our ephemeral public key (-> e)
        let msg1 = buildHandshakeMessage1()
        try await sendFrame(msg1)
        
        // Step 2: Receive server's e, ee, s, es with attestation payload
        let msg2 = try await receiveFrame()
        let (serverPayload, attestationQuote) = try parseHandshakeMessage2(msg2)
        
        // Step 3: Verify attestation if not skipped
        var attestationResult = AttestationResult(
            valid: true,
            attestationType: .unknown,
            errors: [],
            warnings: ["Attestation verification not yet implemented"]
        )
        
        if !config.skipAttestation, let quote = attestationQuote {
            attestationResult = try verifyAttestation(
                quote: quote,
                expectedMeasurements: expectedMeasurements
            )
            
            if !attestationResult.valid {
                throw NoiseError.attestationFailed(attestationResult.errors)
            }
        }
        
        // Step 4: Send our static public key (-> s, se)
        let msg3 = buildHandshakeMessage3()
        try await sendFrame(msg3)
        
        // Step 5: Derive session keys and create session
        let (txKey, rxKey) = deriveSessionKeys()
        
        return NoiseSessionImpl(
            connection: connection,
            txKey: txKey,
            rxKey: rxKey,
            remotePublicKey: remoteStaticPublicKey ?? Data(),
            handshakeHash: computeHandshakeHash(),
            attestation: attestationResult
        )
    }
    
    // MARK: - Crypto Stubs
    // These would use CryptoKit or a Noise Protocol library in production
    
    private func generateKeypair() -> (publicKey: Data, secretKey: Data) {
        // Stub: In production, use Curve25519.KeyAgreement.PrivateKey()
        let secretKey = Data((0..<32).map { _ in UInt8.random(in: 0...255) })
        let publicKey = Data((0..<32).map { _ in UInt8.random(in: 0...255) })
        return (publicKey, secretKey)
    }
    
    private func buildHandshakeMessage1() -> Data {
        // Stub: Return ephemeral public key with Noise framing
        return localEphemeralKeypair?.publicKey ?? Data()
    }
    
    private func parseHandshakeMessage2(_ data: Data) throws -> (payload: Data, attestation: Data?) {
        // Stub: Parse server's response containing:
        // - Server ephemeral public key (32 bytes)
        // - Encrypted server static public key
        // - Encrypted attestation payload
        
        // In production: decrypt and verify using Noise protocol state machine
        
        // Extract attestation quote (simplified - real format has length prefix)
        let attestation: Data? = data.count > 64 ? data.subdata(in: 64..<data.count) : nil
        
        return (data, attestation)
    }
    
    private func buildHandshakeMessage3() -> Data {
        // Stub: Return encrypted static public key
        return localStaticKeypair?.publicKey ?? Data()
    }
    
    private func deriveSessionKeys() -> (tx: Data, rx: Data) {
        // Stub: Derive symmetric keys from handshake
        // In production: use HKDF with handshake transcript hash
        let tx = Data((0..<32).map { _ in UInt8.random(in: 0...255) })
        let rx = Data((0..<32).map { _ in UInt8.random(in: 0...255) })
        return (tx, rx)
    }
    
    private func computeHandshakeHash() -> Data {
        // Stub: Return hash of entire handshake transcript
        return Data((0..<32).map { _ in UInt8.random(in: 0...255) })
    }
    
    // MARK: - Network
    
    private func sendFrame(_ data: Data) async throws {
        guard let connection = connection else {
            throw NoiseError.notConnected
        }
        
        // Length-prefixed framing (4 bytes big-endian)
        var frame = Data(count: 4 + data.count)
        frame[0] = UInt8((data.count >> 24) & 0xFF)
        frame[1] = UInt8((data.count >> 16) & 0xFF)
        frame[2] = UInt8((data.count >> 8) & 0xFF)
        frame[3] = UInt8(data.count & 0xFF)
        frame.replaceSubrange(4..., with: data)
        
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            connection.send(content: frame, completion: .contentProcessed { error in
                if let error = error {
                    continuation.resume(throwing: error)
                } else {
                    continuation.resume()
                }
            })
        }
    }
    
    private func receiveFrame() async throws -> Data {
        guard let connection = connection else {
            throw NoiseError.notConnected
        }
        
        return try await withCheckedThrowingContinuation { continuation in
            // First read the 4-byte length
            connection.receive(minimumIncompleteLength: 4, maximumLength: 4) { data, _, _, error in
                if let error = error {
                    continuation.resume(throwing: error)
                    return
                }
                
                guard let lengthData = data, lengthData.count == 4 else {
                    continuation.resume(throwing: NoiseError.invalidFrame)
                    return
                }
                
                let length = Int(lengthData[0]) << 24 |
                            Int(lengthData[1]) << 16 |
                            Int(lengthData[2]) << 8 |
                            Int(lengthData[3])
                
                // Read the payload
                connection.receive(minimumIncompleteLength: length, maximumLength: length) { data, _, _, error in
                    if let error = error {
                        continuation.resume(throwing: error)
                    } else if let data = data {
                        continuation.resume(returning: data)
                    } else {
                        continuation.resume(throwing: NoiseError.invalidFrame)
                    }
                }
            }
        }
    }
    
    // MARK: - Rekor Integration
    
    private func fetchMeasurementsFromRekor(imageDigest: String) async throws -> ExpectedMeasurements? {
        // Stub: Fetch from Rekor transparency log
        // See AttestationVerifier.swift for implementation details
        return nil
    }
    
    private func verifyAttestation(quote: Data, expectedMeasurements: ExpectedMeasurements?) throws -> AttestationResult {
        // Stub: Delegate to AttestationVerifier
        // See AttestationVerifier.swift for implementation
        return AttestationResult(
            valid: false,
            attestationType: .unknown,
            errors: ["Attestation verification not implemented"],
            warnings: []
        )
    }
}

// MARK: - Session Implementation

private class NoiseSessionImpl: NoiseSession {
    private let connection: NWConnection
    private let txKey: Data
    private let rxKey: Data
    
    let remotePublicKey: Data
    let handshakeHash: Data
    let attestation: AttestationResult
    
    var onData: ((Data) -> Void)?
    
    init(
        connection: NWConnection,
        txKey: Data,
        rxKey: Data,
        remotePublicKey: Data,
        handshakeHash: Data,
        attestation: AttestationResult
    ) {
        self.connection = connection
        self.txKey = txKey
        self.rxKey = rxKey
        self.remotePublicKey = remotePublicKey
        self.handshakeHash = handshakeHash
        self.attestation = attestation
        
        setupReceiveHandler()
    }
    
    func send(_ data: Data) throws {
        // Stub: Encrypt with ChaCha20-Poly1305 using txKey
        let encrypted = encrypt(data, key: txKey)
        
        // Send length-prefixed frame
        var frame = Data(count: 4 + encrypted.count)
        frame[0] = UInt8((encrypted.count >> 24) & 0xFF)
        frame[1] = UInt8((encrypted.count >> 16) & 0xFF)
        frame[2] = UInt8((encrypted.count >> 8) & 0xFF)
        frame[3] = UInt8(encrypted.count & 0xFF)
        frame.replaceSubrange(4..., with: encrypted)
        
        connection.send(content: frame, completion: .contentProcessed { _ in })
    }
    
    func close() {
        connection.cancel()
    }
    
    private func setupReceiveHandler() {
        receiveNextFrame()
    }
    
    private func receiveNextFrame() {
        connection.receive(minimumIncompleteLength: 4, maximumLength: 4) { [weak self] data, _, _, error in
            guard let self = self, error == nil, let lengthData = data else { return }
            
            let length = Int(lengthData[0]) << 24 |
                        Int(lengthData[1]) << 16 |
                        Int(lengthData[2]) << 8 |
                        Int(lengthData[3])
            
            self.connection.receive(minimumIncompleteLength: length, maximumLength: length) { data, _, _, error in
                guard error == nil, let encrypted = data else { return }
                
                // Decrypt with ChaCha20-Poly1305 using rxKey
                let decrypted = self.decrypt(encrypted, key: self.rxKey)
                self.onData?(decrypted)
                
                self.receiveNextFrame()
            }
        }
    }
    
    private func encrypt(_ data: Data, key: Data) -> Data {
        // Stub: Use CryptoKit's ChaChaPoly in production
        return data  // Placeholder
    }
    
    private func decrypt(_ data: Data, key: Data) -> Data {
        // Stub: Use CryptoKit's ChaChaPoly in production
        return data  // Placeholder
    }
}

// MARK: - Supporting Types

public struct ExpectedMeasurements {
    public let snpMeasurement: Data?
    public let tdxMrTd: Data?
    public let tdxRtmr0: Data?
}

public enum NoiseError: Error {
    case notConnected
    case invalidFrame
    case handshakeFailed(String)
    case attestationFailed([String])
    case encryptionError(String)
}

File: client/web/src/noise-client.ts
/**
 * Noise Protocol Client with Attestation Verification
 * 
 * Implements a Noise XX handshake client that:
 * 1. Connects to the server over TCP/TLS
 * 2. Extracts attestation quote from the handshake payload
 * 3. Verifies attestation against expected measurements from Rekor
 * 4. Returns an encrypted duplex stream if verification passes
 */

import { createConnection, Socket } from 'net';
import { connect as tlsConnect, TLSSocket } from 'tls';
import { EventEmitter } from 'events';
import {
  verifyAttestation,
  ExpectedMeasurements,
  VerificationResult,
  hexToBytes
} from './verify.js';
import { fetchExpectedMeasurements, Measurements } from './transparency.js';

// @ts-ignore - noise-handshake is a CommonJS module without types
import Noise from 'noise-handshake';
// @ts-ignore
import Cipher from 'noise-handshake/cipher';

/**
 * Connection options for the Noise client
 */
export interface NoiseClientOptions {
  /** Server hostname */
  host: string;
  /** Server port */
  port: number;
  /** Use TLS as transport (recommended) */
  useTls?: boolean;
  /** Static keypair for the client (generated if not provided) */
  staticKeypair?: { publicKey: Uint8Array; secretKey: Uint8Array };
  /** Expected image digest for fetching measurements from Rekor */
  expectedImageDigest?: string;
  /** Pre-loaded expected measurements (if not fetching from Rekor) */
  expectedMeasurements?: ExpectedMeasurements;
  /** Skip attestation verification (DANGEROUS - for testing only) */
  skipAttestation?: boolean;
  /** Connection timeout in milliseconds */
  timeoutMs?: number;
}

/**
 * Encrypted session after successful handshake
 */
export interface NoiseSession {
  /** Send an encrypted message */
  send(data: Uint8Array): void;
  /** Receive handler - call to process incoming data */
  onData(handler: (decrypted: Uint8Array) => void): void;
  /** Close the session */
  close(): void;
  /** Server's static public key */
  remotePublicKey: Uint8Array;
  /** Handshake hash (for channel binding) */
  handshakeHash: Uint8Array;
  /** Attestation verification result */
  attestation: VerificationResult;
}

/**
 * Attestation payload embedded in handshake
 * 
 * The server sends this in the second handshake message (e -> e, ee, s, es)
 * The payload contains:
 * - 4-byte length prefix (big-endian)
 * - Attestation quote bytes
 */
interface AttestationPayload {
  quote: Uint8Array;
}

/**
 * Parse attestation payload from handshake message
 */
function parseAttestationPayload(payload: Uint8Array): AttestationPayload | null {
  if (payload.length < 4) {
    return null;
  }
  
  const view = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
  const quoteLength = view.getUint32(0, false);  // big-endian
  
  if (payload.length < 4 + quoteLength) {
    return null;
  }
  
  return {
    quote: payload.slice(4, 4 + quoteLength)
  };
}

/**
 * NoiseClient - Establishes attested encrypted connections
 */
export class NoiseClient extends EventEmitter {
  private socket: Socket | TLSSocket | null = null;
  private noise: typeof Noise | null = null;
  private sendCipher: typeof Cipher | null = null;
  private recvCipher: typeof Cipher | null = null;
  private options: NoiseClientOptions;
  private connected = false;
  private dataHandlers: ((data: Uint8Array) => void)[] = [];
  private receiveBuffer: Buffer = Buffer.alloc(0);
  
  constructor(options: NoiseClientOptions) {
    super();
    this.options = {
      useTls: true,
      skipAttestation: false,
      timeoutMs: 30000,
      ...options
    };
  }
  
  /**
   * Connect to the server and perform attested handshake
   */
  async connect(): Promise<NoiseSession> {
    // Fetch expected measurements from Rekor if image digest provided
    let expectedMeasurements = this.options.expectedMeasurements;
    
    if (!expectedMeasurements && this.options.expectedImageDigest) {
      console.log('Fetching expected measurements from Rekor...');
      const measurements = await fetchExpectedMeasurements(
        this.options.expectedImageDigest
      );
      
      if (measurements) {
        expectedMeasurements = {
          snpMeasurement: measurements.snpMeasurement 
            ? hexToBytes(measurements.snpMeasurement)
            : undefined,
          tdxMrTd: measurements.tdxMrTd
            ? hexToBytes(measurements.tdxMrTd)
            : undefined,
          tdxRtmr0: measurements.tdxRtmr0
            ? hexToBytes(measurements.tdxRtmr0)
            : undefined
        };
        console.log(`Found measurements from Rekor (log index: ${measurements.logIndex})`);
      } else {
        console.warn('No measurements found in Rekor for image digest');
      }
    }
    
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.cleanup();
        reject(new Error('Connection timeout'));
      }, this.options.timeoutMs);
      
      const onError = (err: Error) => {
        clearTimeout(timeout);
        this.cleanup();
        reject(err);
      };
      
      // Initialize Noise handshake state (XX pattern, initiator)
      this.noise = new Noise('XX', true, this.options.staticKeypair);
      this.noise.initialise(Buffer.alloc(0));  // empty prologue
      
      // Establish TCP/TLS connection
      const connectFn = () => {
        if (this.options.useTls) {
          this.socket = tlsConnect({
            host: this.options.host,
            port: this.options.port,
            rejectUnauthorized: false  // We verify via attestation, not TLS CA
          });
        } else {
          this.socket = createConnection({
            host: this.options.host,
            port: this.options.port
          });
        }
        
        this.socket.once('error', onError);
        
        this.socket.once('connect', () => {
          this.performHandshake(expectedMeasurements)
            .then(session => {
              clearTimeout(timeout);
              resolve(session);
            })
            .catch(err => {
              clearTimeout(timeout);
              this.cleanup();
              reject(err);
            });
        });
      };
      
      connectFn();
    });
  }
  
  /**
   * Perform the Noise XX handshake with attestation verification
   */
  private async performHandshake(
    expectedMeasurements?: ExpectedMeasurements
  ): Promise<NoiseSession> {
    if (!this.socket || !this.noise) {
      throw new Error('Not connected');
    }
    
    // XX handshake pattern:
    // -> e                    (initiator sends ephemeral)
    // <- e, ee, s, es         (responder sends ephemeral + static + attestation)
    // -> s, se                (initiator sends static)
    
    return new Promise((resolve, reject) => {
      let handshakeStep = 0;
      let attestationResult: VerificationResult | null = null;
      
      // Step 1: Send initiator's ephemeral key
      const msg1 = this.noise!.send();
      this.sendFrame(msg1);
      handshakeStep = 1;
      
      // Handle incoming handshake messages
      const onData = (data: Buffer) => {
        this.receiveBuffer = Buffer.concat([this.receiveBuffer, data]);
        
        try {
          // Try to parse a complete frame
          const frame = this.readFrame();
          if (!frame) return;  // Need more data
          
          if (handshakeStep === 1) {
            // Step 2: Receive responder's e, ee, s, es with attestation payload
            const payload = this.noise!.recv(frame);
            
            // Parse attestation from payload
            const attestationPayload = parseAttestationPayload(payload);
            
            if (attestationPayload && !this.options.skipAttestation) {
              // Verify attestation
              attestationResult = verifyAttestation(
                attestationPayload.quote,
                expectedMeasurements || {}
              );
              
              if (!attestationResult.valid) {
                throw new Error(
                  `Attestation verification failed: ${attestationResult.errors.join(', ')}`
                );
              }
              
              console.log('Attestation verified successfully');
              for (const warning of attestationResult.warnings) {
                console.warn(`Attestation warning: ${warning}`);
              }
            } else if (!this.options.skipAttestation) {
              console.warn('No attestation payload in handshake');
            }
            
            // Step 3: Send initiator's static key
            const msg3 = this.noise!.send();
            this.sendFrame(msg3);
            handshakeStep = 2;
            
          } else if (handshakeStep === 2) {
            // Optional: Handle any final handshake message from server
            // XX pattern completes after msg3, but server might send confirmation
            
            if (this.noise!.complete) {
              this.finalizeHandshake(attestationResult, resolve);
              this.socket!.off('data', onData);
            }
          }
          
          // Check if handshake is complete
          if (this.noise!.complete && handshakeStep >= 2) {
            this.finalizeHandshake(attestationResult, resolve);
            this.socket!.off('data', onData);
          }
        } catch (err) {
          this.socket!.off('data', onData);
          reject(err);
        }
      };
      
      this.socket!.on("data", onData);
    });
  }
  
  /**
   * Finalize the handshake and create the session
   */
  private finalizeHandshake(
    attestationResult: VerificationResult | null,
    resolve: (session: NoiseSession) => void
  ): void {
    // Create cipher instances for encryption/decryption
    this.sendCipher = new Cipher(this.noise!.tx);
    this.recvCipher = new Cipher(this.noise!.rx);
    this.connected = true;
    
    // Set up data handler for post-handshake messages
    this.socket!.on('data', (data: Buffer) => {
      this.receiveBuffer = Buffer.concat([this.receiveBuffer, data]);
      
      let frame = this.readFrame();
      while (frame) {
        try {
          const decrypted = this.recvCipher!.decrypt(frame);
          for (const handler of this.dataHandlers) {
            handler(new Uint8Array(decrypted));
          }
        } catch (err) {
          this.emit('error', err);
        }
        frame = this.readFrame();
      }
    });
    
    const session: NoiseSession = {
      send: (data: Uint8Array) => {
        if (!this.connected || !this.sendCipher) {
          throw new Error('Session not established');
        }
        const encrypted = this.sendCipher.encrypt(Buffer.from(data));
        this.sendFrame(encrypted);
      },
      onData: (handler: (decrypted: Uint8Array) => void) => {
        this.dataHandlers.push(handler);
      },
      close: () => {
        this.cleanup();
      },
      remotePublicKey: this.noise!.rs,
      handshakeHash: this.noise!.hash,
      attestation: attestationResult || {
        valid: this.options.skipAttestation || false,
        attestationType: 'unknown',
        errors: this.options.skipAttestation ? [] : ['No attestation performed'],
        warnings: this.options.skipAttestation ? ['Attestation skipped'] : []
      }
    };
    
    resolve(session);
  }
  
  /**
   * Send a length-prefixed frame
   */
  private sendFrame(data: Buffer | Uint8Array): void {
    if (!this.socket) throw new Error('Not connected');
    
    const frame = Buffer.alloc(4 + data.length);
    frame.writeUInt32BE(data.length, 0);
    Buffer.from(data).copy(frame, 4);
    this.socket.write(frame);
  }
  
  /**
   * Read a complete frame from the receive buffer
   */
  private readFrame(): Buffer | null {
    if (this.receiveBuffer.length < 4) {
      return null;
    }
    
    const length = this.receiveBuffer.readUInt32BE(0);
    
    if (this.receiveBuffer.length < 4 + length) {
      return null;
    }
    
    const frame = this.receiveBuffer.slice(4, 4 + length);
    this.receiveBuffer = this.receiveBuffer.slice(4 + length);
    
    return frame;
  }
  
  /**
   * Clean up resources
   */
  private cleanup(): void {
    this.connected = false;
    if (this.socket) {
      this.socket.destroy();
      this.socket = null;
    }
    this.noise = null;
    this.sendCipher = null;
    this.recvCipher = null;
    this.dataHandlers = [];
  }
}

/**
 * Create a Noise client and connect
 * 
 * Convenience function that creates a client and performs the handshake.
 * 
 * @param options - Connection options
 * @returns Encrypted session with attestation verification
 */
export async function createNoiseClient(
  options: NoiseClientOptions
): Promise<NoiseSession> {
  const client = new NoiseClient(options);
  return client.connect();
}
